File: support-files/src/pipeline-plugin/bksdk/src/main/java/com/tencent/bk/devops/atom/utils/http/OkHttpUtils.java
package com.tencent.bk.devops.atom.utils.http;

import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * http请求工具类
 */
public class OkHttpUtils {

    private final static String CONTENT_TYPE_JSON = "application/json; charset=utf-8";

    private final static Logger logger = LoggerFactory.getLogger(OkHttpUtils.class);


    private static OkHttpClient createClient(long connectTimeout, long writeTimeout, long readTimeout) {
        long finalConnectTimeout = 5L;
        long finalWriteTimeout = 60L;
        long finalReadTimeout = 60L;
        OkHttpClient.Builder builder = new okhttp3.OkHttpClient.Builder();
        if (connectTimeout > 0)
            finalConnectTimeout = connectTimeout;
        if (writeTimeout > 0)
            finalWriteTimeout = writeTimeout;
        if (readTimeout > 0)
            finalReadTimeout = readTimeout;
        builder.writeTimeout(finalConnectTimeout, TimeUnit.SECONDS);
        builder.writeTimeout(finalWriteTimeout, TimeUnit.SECONDS);
        builder.readTimeout(finalReadTimeout, TimeUnit.SECONDS);
        return builder.build();
    }


    private static Request.Builder getBuilder(String url, Map<String, String> headers) {
        Request.Builder builder = new Request.Builder();
        builder.url(url);
        if (null != headers) {
            builder.headers(Headers.of(headers));
        }
        return builder;
    }

    /**
     * http get方式请求，返回json格式响应报文
     *
     * @param url 请求路径
     * @return json格式响应报文
     */
    public static String doGet(String url) {
        return doGet(url, null);
    }

    /**
     * http get方式请求，返回json格式响应报文
     *
     * @param url            请求路径
     * @param connectTimeout 连接超时时间
     * @param writeTimeout   写超时时间
     * @param readTimeout    读超时时间
     * @return json格式响应报文
     */
    public static String doGet(String url, long connectTimeout, long writeTimeout, long readTimeout) {
        return doGet(url, null, connectTimeout, writeTimeout, readTimeout);
    }

    /**
     * http get方式请求，返回json格式响应报文
     *
     * @param url     请求路径
     * @param headers 请求头
     * @return json格式响应报文
     */
    public static String doGet(String url, Map<String, String> headers) {
        return doGet(url, headers, -1, -1, -1);
    }

    /**
     * http get方式请求，返回json格式响应报文
     *
     * @param url            请求路径
     * @param headers        请求头
     * @param connectTimeout 连接超时时间
     * @param writeTimeout   写超时时间
     * @param readTimeout    读超时时间
     * @return json格式响应报文
     */
    public static String doGet(String url, Map<String, String> headers, long connectTimeout, long writeTimeout, long readTimeout) {
        Request.Builder builder = getBuilder(url, headers);
        Request request = builder.get().build();
        return doHttp(request, connectTimeout, writeTimeout, readTimeout);
    }

    /**
     * http post方式请求，返回json格式响应报文
     *
     * @param url       请求路径
     * @param jsonParam json格式参数
     * @return json格式响应报文
     */
    public static String doPost(String url, String jsonParam) {
        return doPost(url, jsonParam, null);
    }

    /**
     * http post方式请求，返回json格式响应报文
     *
     * @param url            请求路径
     * @param jsonParam      json格式参数
     * @param connectTimeout 连接超时时间
     * @param writeTimeout   写超时时间
     * @param readTimeout    读超时时间
     * @return json格式响应报文
     */
    public static String doPost(String url, String jsonParam, long connectTimeout, long writeTimeout, long readTimeout) {
        return doPost(url, jsonParam, null, connectTimeout, writeTimeout, readTimeout);
    }

    /**
     * http post方式请求，返回json格式响应报文
     *
     * @param url       请求路径
     * @param jsonParam json格式参数
     * @param headers   请求头
     * @return json格式响应报文
     */
    public static String doPost(String url, String jsonParam, Map<String, String> headers) {
        return doPost(url, jsonParam, headers, -1, -1, -1);
    }

    /**
     * http post方式请求，返回json格式响应报文
     *
     * @param url            请求路径
     * @param jsonParam      json格式参数
     * @param headers        请求头
     * @param connectTimeout 连接超时时间
     * @param writeTimeout   写超时时间
     * @param readTimeout    读超时时间
     * @return json格式响应报文
     */
    public static String doPost(String url, String jsonParam, Map<String, String> headers, long connectTimeout, long writeTimeout, long readTimeout) {
        Request.Builder builder = getBuilder(url, headers);
        RequestBody body = RequestBody.create(MediaType.parse(CONTENT_TYPE_JSON), jsonParam);
        Request request = builder.post(body).build();
        return doHttp(request, connectTimeout, writeTimeout, readTimeout);
    }

    /**
     * http put方式请求，返回json格式响应报文
     *
     * @param url       请求路径
     * @param jsonParam json格式参数
     * @return json格式响应报文
     */
    public static String doPut(String url, String jsonParam) {
        return doPut(url, jsonParam, null);
    }

    /**
     * http put方式请求，返回json格式响应报文
     *
     * @param url            请求路径
     * @param jsonParam      json格式参数
     * @param connectTimeout 连接超时时间
     * @param writeTimeout   写超时时间
     * @param readTimeout    读超时时间
     * @return json格式响应报文
     */
    public static String doPut(String url, String jsonParam, long connectTimeout, long writeTimeout, long readTimeout) {
        return doPut(url, jsonParam, null, connectTimeout, writeTimeout, readTimeout);
    }

    /**
     * http put方式请求，返回json格式响应报文
     *
     * @param url       请求路径
     * @param jsonParam json格式参数
     * @param headers   请求头
     * @return json格式响应报文
     */
    public static String doPut(String url, String jsonParam, Map<String, String> headers) {
        return doPut(url, jsonParam, headers, -1, -1, -1);
    }

    /**
     * http put方式请求，返回json格式响应报文
     *
     * @param url            请求路径
     * @param jsonParam      json格式参数
     * @param headers        请求头
     * @param connectTimeout 连接超时时间
     * @param writeTimeout   写超时时间
     * @param readTimeout    读超时时间
     * @return json格式响应报文
     */
    public static String doPut(String url, String jsonParam, Map<String, String> headers, long connectTimeout, long writeTimeout, long readTimeout) {
        Request.Builder builder = getBuilder(url, headers);
        RequestBody body = RequestBody.create(MediaType.parse(CONTENT_TYPE_JSON), jsonParam);
        Request request = builder.put(body).build();
        return doHttp(request, connectTimeout, writeTimeout, readTimeout);
    }

    /**
     * http delete方式请求，返回json格式响应报文
     *
     * @param url 请求路径
     * @return json格式响应报文
     */
    public static String doDelete(String url) {
        return doDelete(url, null);
    }

    /**
     * http delete方式请求，返回json格式响应报文
     *
     * @param url            请求路径
     * @param connectTimeout 连接超时时间
     * @param writeTimeout   写超时时间
     * @param readTimeout    读超时时间
     * @return json格式响应报文
     */
    public static String doDelete(String url, long connectTimeout, long writeTimeout, long readTimeout) {
        return doDelete(url, null, connectTimeout, writeTimeout, readTimeout);
    }

    /**
     * http delete方式请求，返回json格式响应报文
     *
     * @param url     请求路径
     * @param headers 请求头
     * @return json格式响应报文
     */
    public static String doDelete(String url, Map<String, String> headers) {
        return doDelete(url, headers, -1, -1, -1);
    }

    /**
     * http delete方式请求，返回json格式响应报文
     *
     * @param url            请求路径
     * @param headers        请求头
     * @param connectTimeout 连接超时时间
     * @param writeTimeout   写超时时间
     * @param readTimeout    读超时时间
     * @return json格式响应报文
     */
    public static String doDelete(String url, Map<String, String> headers, long connectTimeout, long writeTimeout, long readTimeout) {
        Request.Builder builder = getBuilder(url, headers);
        Request request = builder.delete().build();
        return doHttp(request, connectTimeout, writeTimeout, readTimeout);
    }

    public static String doHttp(Request request, long connectTimeout, long writeTimeout, long readTimeout) {
        OkHttpClient httpClient = createClient(connectTimeout, writeTimeout, readTimeout);
        Response response = null;
        String responseContent = null;
        try {
            response = httpClient.newCall(request).execute();
            assert response.body() != null;
            responseContent = response.body().string();
        } catch (IOException e) {
            logger.error("http send  throw Exception", e);
        } finally {
            if (response != null) {
                assert response.body() != null;
                response.body().close();
            }
        }
        if (response != null && !response.isSuccessful()) {
            logger.error("Fail to request(" + request + ") with code " + response.code()
                    + " , message " + response.message() + " and response" + responseContent);
        }
        return responseContent;
    }

}

File: src/backend/codecc/core/common/common-util/src/main/kotlin/com/tencent/devops/common/web/utils/OkhttpUtils.kt
/*
 * Tencent is pleased to support the open source community by making BK-CODECC 蓝鲸代码检查平台 available.
 *
 * Copyright (C) 2019 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * BK-CODECC 蓝鲸代码检查平台 is licensed under the MIT license.
 *
 * A copy of the MIT License is included in this file.
 *
 *
 * Terms of the MIT License:
 * ---------------------------------------------------
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy,
 * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
 * NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.tencent.devops.common.util

import com.tencent.devops.common.api.exception.CodeCCException
import com.tencent.devops.common.constant.CommonMessageCode
import okhttp3.MediaType
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.TimeUnit

object OkhttpUtils {

    private val logger = LoggerFactory.getLogger(OkhttpUtils::class.java)

    val okHttpClient = okhttp3.OkHttpClient.Builder()
            .connectTimeout(60L, TimeUnit.SECONDS)
            .readTimeout(30L, TimeUnit.MINUTES)
            .writeTimeout(30L, TimeUnit.MINUTES)
            .build()!!

    fun doGet(url: String, headers: Map<String, String> = mapOf()): String {
        val requestBuilder = Request.Builder()
                .url(url)
                .get()
        if (headers.isNotEmpty()) {
            headers.forEach { key, value ->
                requestBuilder.addHeader(key, value)
            }
        }
        val request = requestBuilder.build()

        okHttpClient.newCall(request).execute().use { response ->
            val responseContent = response.body()!!.string()
            if (!response.isSuccessful) {
                logger.warn("request failed, url:$url message: ${response.message()}")
                throw CodeCCException(CommonMessageCode.THIRD_PARTY_SYSTEM_FAIL)
            }
            return responseContent
        }
    }

    fun doGet(url: String, parameters: Map<String, Any>, headers: Map<String, String>): String {
        var parameterUrl = StringBuffer()
        if (!parameters.isNullOrEmpty()) {
            parameters.forEach { parameterUrl.append("&${it.key}=${it.value}") }
        }
        return doGet(url + parameterUrl, headers)
    }

    fun doHttp(request: Request): Response {
        return okHttpClient.newCall(request).execute()
    }


    fun doHttpPost(url: String, body: String, headers: Map<String, String> = mapOf()): String {
        val requestBuilder = Request.Builder()
                .url(url)
                .post(RequestBody.create(
                        MediaType.parse("application/json; charset=utf-8"), body))
        if (headers.isNotEmpty()) {
            headers.forEach { key, value ->
                requestBuilder.addHeader(key, value)
            }
        }
        val request = requestBuilder.build()
        val client = okHttpClient.newBuilder().build()
        client.newCall(request).execute().use { response ->
            val responseContent = response.body()!!.string()
            if (!response.isSuccessful) {
                logger.warn("request failed, url: $url requestBody: $body message: ${response.message()}, content: $responseContent")
                throw CodeCCException(CommonMessageCode.THIRD_PARTY_SYSTEM_FAIL)
            }
            return responseContent
        }
    }


    fun doHttpDelete(url: String, body: String, headers: Map<String, String> = mapOf()): String {
        val requestBuilder = Request.Builder()
                .url(url)
                .delete(RequestBody.create(
                        MediaType.parse("application/json; charset=utf-8"), body))
        if (headers.isNotEmpty()) {
            headers.forEach { key, value ->
                requestBuilder.addHeader(key, value)
            }
        }
        val request = requestBuilder.build()
        val client = okHttpClient.newBuilder().build()
        client.newCall(request).execute().use { response ->
            val responseContent = response.body()!!.string()
            if (!response.isSuccessful) {
                logger.warn("request failed, message: ${response.message()}")
                throw CodeCCException(CommonMessageCode.THIRD_PARTY_SYSTEM_FAIL)
            }
            return responseContent
        }
    }


    fun doHttpPut(url: String, body: String, headers: Map<String, String> = mapOf()): String {
        val requestBuilder = Request.Builder()
                .url(url)
                .put(RequestBody.create(
                        MediaType.parse("application/json; charset=utf-8"), body))
        if (headers.isNotEmpty()) {
            headers.forEach { key, value ->
                requestBuilder.addHeader(key, value)
            }
        }
        val request = requestBuilder.build()
        val client = okHttpClient.newBuilder().build()
        client.newCall(request).execute().use { response ->
            val responseContent = response.body()!!.string()
            if (!response.isSuccessful) {
                logger.warn("request failed, message: ${response.message()}")
                throw CodeCCException(CommonMessageCode.THIRD_PARTY_SYSTEM_FAIL)
            }
            return responseContent
        }
    }

    fun doFileStreamPut(url: String, file: File, headers: Map<String, String> = mapOf()): String {
        val requestBuilder = Request.Builder()
            .url(url)
            .put(RequestBody.create(
                MediaType.parse("application/octet-stream; charset=utf-8"), file))
        if (headers.isNotEmpty()) {
            headers.forEach { key, value ->
                requestBuilder.addHeader(key, value)
            }
        }
        val request = requestBuilder.build()
        val client = okHttpClient.newBuilder().build()
        client.newCall(request).execute().use { response ->
            val responseContent = response.body()!!.string()
            if (!response.isSuccessful) {
                logger.warn("request failed, message: ${response.message()}")
                throw CodeCCException(CommonMessageCode.THIRD_PARTY_SYSTEM_FAIL)
            }
            return responseContent
        }
    }

    fun downloadFile(url: String, destPath: File) {
        downloadFile(url, destPath, mapOf())
    }

    // NOCC:NestedBlockDepth(设计如此:)
    fun downloadFile(url: String, destPath: File, headers: Map<String, String> = mapOf()) {
        val requestBuilder = Request.Builder()
            .url(url)
            .get()

        if (headers.isNotEmpty()) {
            headers.forEach { key, value ->
                requestBuilder.addHeader(key, value)
            }
        }
        val request = requestBuilder.build()
        okHttpClient.newCall(request).execute().use { response ->
            if (response.code() == HttpStatus.NOT_FOUND.value()) {
                logger.warn("The file $url is not exist")
                throw RuntimeException("文件不存在")
            }
            if (!response.isSuccessful) {
                logger.warn("fail to download the file from $url because of" +
                        " ${response.message()} and code ${response.code()}")
                throw RuntimeException("获取文件失败")
            }
            if (!destPath.parentFile.exists()) destPath.parentFile.mkdirs()
            val buf = ByteArray(4096)
            response.body()!!.byteStream().use { bs ->
                var len = bs.read(buf)
                FileOutputStream(destPath).use { fos ->
                    while (len != -1) {
                        fos.write(buf, 0, len)
                        len = bs.read(buf)
                    }
                }
            }
        }
    }

    fun downloadFile(response: Response, destPath: File) {
        if (response.code() == HttpStatus.NOT_MODIFIED.value()) {
            logger.info("file is newest, do not download to $destPath")
            return
        }
        if (!response.isSuccessful) {
            logger.warn("fail to download the file because of ${response.message()} and code ${response.code()}")
            throw RuntimeException("获取文件失败")
        }
        if (!destPath.parentFile.exists()) destPath.parentFile.mkdirs()
        val buf = ByteArray(4096)
        response.body()!!.byteStream().use { bs ->
            var len = bs.read(buf)
            FileOutputStream(destPath).use { fos ->
                while (len != -1) {
                    fos.write(buf, 0, len)
                    len = bs.read(buf)
                }
            }
        }
    }
}